File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.c
#include <stdlib.h>

#define MAX(a, b) ((a) > (b) ? (a) : (b))
#define MIN(a, b) ((a) < (b) ? (a) : (b))

int firstStableIndex(int* nums, int numsSize, int k) {
    if (numsSize == 0) return -1;

    int* prefixMax = (int*)malloc(numsSize * sizeof(int));
    int* suffMin = (int*)malloc(numsSize * sizeof(int));

    // Compute prefix maximums
    prefixMax[0] = nums[0];
    for (int i = 1; i < numsSize; i++) {
        prefixMax[i] = MAX(prefixMax[i - 1], nums[i]);
    }

    // Compute suffix minimums
    suffMin[numsSize - 1] = nums[numsSize - 1];
    for (int i = numsSize - 2; i >= 0; i--) {
        suffMin[i] = MIN(suffMin[i + 1], nums[i]);
    }

    // Find smallest stable index
    int answer = -1;
    for (int i = 0; i < numsSize; i++) {
        if (prefixMax[i] - suffMin[i] <= k) {
            answer = i;
            break;
        }
    }

    // Free dynamically allocated arrays to prevent memory leaks
    free(prefixMax);
    free(suffMin);

    return answer;
}

File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        n = len(nums)
        if n == 0:
            return -1

        prefix_max = [0] * n
        suff_min = [0] * n

        # Compute prefix maximums
        prefix_max[0] = nums[0]
        for i in range(1, n):
            prefix_max[i] = max(prefix_max[i - 1], nums[i])

        # Compute suffix minimums
        suff_min[n - 1] = nums[n - 1]
        for i in range(n - 2, -1, -1):
            suff_min[i] = min(suff_min[i + 1], nums[i])

        # Find smallest stable index
        for i in range(n):
            if prefix_max[i] - suff_min[i] <= k:
                return i

        return -1
